File: backend/models/sales.py
"""
Sales Model - Process sales transactions with advanced features
"""
import sqlite3
import json
from datetime import datetime
from typing import Dict, List, Optional, Tuple
from decimal import Decimal, ROUND_HALF_UP

class SalesModel:
    def __init__(self, db_path: str = "pos_database.db"):
        self.db_path = db_path
        self.init_database()
    
    def init_database(self):
        """Initialize sales tables"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Sales transactions table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS sales (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                transaction_id TEXT UNIQUE NOT NULL,
                customer_id INTEGER,
                employee_id INTEGER NOT NULL,
                subtotal DECIMAL(10,2) NOT NULL,
                tax_amount DECIMAL(10,2) NOT NULL,
                discount_amount DECIMAL(10,2) DEFAULT 0,
                total_amount DECIMAL(10,2) NOT NULL,
                payment_method TEXT NOT NULL,
                payment_status TEXT DEFAULT 'completed',
                items JSON NOT NULL,
                receipt_data JSON,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                FOREIGN KEY (customer_id) REFERENCES customers(id),
                FOREIGN KEY (employee_id) REFERENCES employees(id)
            )
        ''')
        
        # Cash drawer table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS cash_drawer (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                employee_id INTEGER NOT NULL,
                opening_amount DECIMAL(10,2) NOT NULL,
                closing_amount DECIMAL(10,2),
                expected_amount DECIMAL(10,2),
                difference DECIMAL(10,2),
                status TEXT DEFAULT 'open',
                opened_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                closed_at TIMESTAMP,
                FOREIGN KEY (employee_id) REFERENCES employees(id)
            )
        ''')
        
        # Tax rates table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS tax_rates (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                name TEXT NOT NULL,
                rate DECIMAL(5,4) NOT NULL,
                is_active BOOLEAN DEFAULT 1,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        ''')
        
        # Insert default tax rate
        cursor.execute('''
            INSERT OR IGNORE INTO tax_rates (name, rate) 
            VALUES ('Standard Tax', 0.0875)
        ''')
        
        conn.commit()
        conn.close()
    
    def process_sale(self, sale_data: Dict) -> Dict:
        """Process a complete sales transaction"""
        try:
            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()
            
            # Calculate totals
            subtotal = self._calculate_subtotal(sale_data['items'])
            tax_amount = self._calculate_tax(subtotal, sale_data.get('tax_rate', 0.0875))
            discount_amount = self._calculate_discount(subtotal, sale_data.get('discount'))
            total_amount = subtotal + tax_amount - discount_amount
            
            # Generate transaction ID
            transaction_id = self._generate_transaction_id()
            
            # Insert sale record
            cursor.execute('''
                INSERT INTO sales (
                    transaction_id, customer_id, employee_id, subtotal, 
                    tax_amount, discount_amount, total_amount, payment_method, 
                    items, receipt_data
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            ''', (
                transaction_id,
                sale_data.get('customer_id'),
                sale_data['employee_id'],
                float(subtotal),
                float(tax_amount),
                float(discount_amount),
                float(total_amount),
                sale_data['payment_method'],
                json.dumps(sale_data['items']),
                json.dumps(self._generate_receipt_data(sale_data, transaction_id, subtotal, tax_amount, discount_amount, total_amount))
            ))
            
            # Update inventory if items are physical products
            self._update_inventory_after_sale(sale_data['items'])
            
            conn.commit()
            sale_id = cursor.lastrowid
            
            conn.close()
            
            return {
                'success': True,
                'transaction_id': transaction_id,
                'sale_id': sale_id,
                'total_amount': float(total_amount),
                'receipt_data': self._generate_receipt_data(sale_data, transaction_id, subtotal, tax_amount, discount_amount, total_amount)
            }
            
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    def void_transaction(self, transaction_id: str, employee_id: int, reason: str = "") -> Dict:
        """Void a completed transaction"""
        try:
            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()
            
            # Get the original sale
            cursor.execute('''
                SELECT * FROM sales WHERE transaction_id = ? AND payment_status = 'completed'
            ''', (transaction_id,))
            
            sale = cursor.fetchone()
            if not sale:
                return {'success': False, 'error': 'Transaction not found or already voided'}
            
            # Restore inventory
            items = json.loads(sale[10])  # items column
            self._restore_inventory_after_void(items)
            
            # Mark as voided
            cursor.execute('''
                UPDATE sales 
                SET payment_status = 'voided', updated_at = CURRENT_TIMESTAMP
                WHERE transaction_id = ?
            ''', (transaction_id,))
            
            conn.commit()
            conn.close()
            
            return {'success': True, 'message': 'Transaction voided successfully'}
            
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    def apply_discount(self, transaction_id: str, discount_type: str, discount_value: float) -> Dict:
        """Apply discount to a transaction"""
        try:
            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()
            
            # Get current sale
            cursor.execute('SELECT * FROM sales WHERE transaction_id = ?', (transaction_id,))
            sale = cursor.fetchone()
            
            if not sale:
                return {'success': False, 'error': 'Transaction not found'}
            
            subtotal = float(sale[4])  # subtotal column
            current_discount = float(sale[6])  # discount_amount column
            
            # Calculate new discount
            if discount_type == 'percentage':
                new_discount = subtotal * (discount_value / 100)
            else:  # fixed amount
                new_discount = discount_value
            
            # Ensure discount doesn't exceed subtotal
            new_discount = min(new_discount, subtotal)
            
            # Recalculate total
            tax_amount = float(sale[5])
            new_total = subtotal + tax_amount - new_discount
            
            # Update sale
            cursor.execute('''
                UPDATE sales 
                SET discount_amount = ?, total_amount = ?, updated_at = CURRENT_TIMESTAMP
                WHERE transaction_id = ?
            ''', (float(new_discount), float(new_total), transaction_id))
            
            conn.commit()
            conn.close()
            
            return {
                'success': True,
                'new_discount': float(new_discount),
                'new_total': float(new_total)
            }
            
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    def calculate_tax(self, subtotal: float, tax_rate: float = 0.0875) -> float:
        """Calculate tax amount"""
        return float(Decimal(str(subtotal * tax_rate)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
    
    def get_sales_summary(self, start_date: str, end_date: str) -> Dict:
        """Get sales summary for date range"""
        try:
            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()
            
            cursor.execute('''
                SELECT 
                    COUNT(*) as total_transactions,
                    SUM(total_amount) as total_revenue,
                    AVG(total_amount) as average_transaction,
                    SUM(tax_amount) as total_tax,
                    SUM(discount_amount) as total_discounts
                FROM sales 
                WHERE created_at BETWEEN ? AND ? AND payment_status = 'completed'
            ''', (start_date, end_date))
            
            summary = cursor.fetchone()
            conn.close()
            
            return {
                'success': True,
                'summary': {
                    'total_transactions': summary[0] or 0,
                    'total_revenue': float(summary[1] or 0),
                    'average_transaction': float(summary[2] or 0),
                    'total_tax': float(summary[3] or 0),
                    'total_discounts': float(summary[4] or 0)
                }
            }
            
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    def _calculate_subtotal(self, items: List[Dict]) -> float:
        """Calculate subtotal from items"""
        subtotal = 0
        for item in items:
            subtotal += item['price'] * item['quantity']
        return float(Decimal(str(subtotal)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
    
    def _calculate_tax(self, subtotal: float, tax_rate: float) -> float:
        """Calculate tax amount"""
        return self.calculate_tax(subtotal, tax_rate)
    
    def _calculate_discount(self, subtotal: float, discount: Optional[Dict]) -> float:
        """Calculate discount amount"""
        if not discount:
            return 0.0
        
        if discount['type'] == 'percentage':
            return float(Decimal(str(subtotal * discount['value'] / 100)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
        else:
            return min(discount['value'], subtotal)
    
    def _generate_transaction_id(self) -> str:
        """Generate unique transaction ID"""
        timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
        import random
        random_suffix = random.randint(1000, 9999)
        return f"TXN{timestamp}{random_suffix}"
    
    def _generate_receipt_data(self, sale_data: Dict, transaction_id: str, 
                              subtotal: float, tax_amount: float, 
                              discount_amount: float, total_amount: float) -> Dict:
        """Generate receipt data"""
        return {
            'transaction_id': transaction_id,
            'date': datetime.now().isoformat(),
            'items': sale_data['items'],
            'subtotal': subtotal,
            'tax_amount': tax_amount,
            'discount_amount': discount_amount,
            'total_amount': total_amount,
            'payment_method': sale_data['payment_method'],
            'employee_id': sale_data['employee_id'],
            'customer_id': sale_data.get('customer_id')
        }
    
    def _update_inventory_after_sale(self, items: List[Dict]):
        """Update inventory quantities after sale"""
        # This would integrate with inventory model
        # For now, just a placeholder
        pass
    
    def _restore_inventory_after_void(self, items: List[Dict]):
        """Restore inventory quantities after void"""
        # This would integrate with inventory model
        # For now, just a placeholder
        pass

File: backend/routes/sales.js
const express = require('express');
const router = express.Router();

// Import models
const SalesModel = require('../models/sales');
const InventoryModel = require('../models/inventory');
const CustomerModel = require('../models/customer');
const AuthenticationManager = require('../utils/authentication');

// Initialize models
const salesModel = new SalesModel();
const inventoryModel = new InventoryModel();
const customerModel = new CustomerModel();
const authManager = new AuthenticationManager();

// Middleware to verify authentication
const authenticateUser = async (req, res, next) => {
  try {
    const session_token = req.cookies.session_token || req.headers.authorization?.replace('Bearer ', '');
    
    if (!session_token) {
      return res.status(401).json({
        success: false,
        message: 'Authentication required'
      });
    }

    const session = await authManager.validate_session(session_token);
    if (!session) {
      return res.status(401).json({
        success: false,
        message: 'Invalid or expired session'
      });
    }

    req.user = session;
    next();
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Authentication error'
    });
  }
};

// Process a new sale
router.post('/process', authenticateUser, async (req, res) => {
  try {
    const sale_data = {
      ...req.body,
      employee_id: req.user.user_id
    };

    // Validate required fields
    if (!sale_data.items || !Array.isArray(sale_data.items) || sale_data.items.length === 0) {
      return res.status(400).json({
        success: false,
        message: 'Items are required for a sale'
      });
    }

    if (!sale_data.payment_method) {
      return res.status(400).json({
        success: false,
        message: 'Payment method is required'
      });
    }

    // Process the sale
    const result = await salesModel.process_sale(sale_data);

    if (result.success) {
      // Update customer purchase history if customer_id provided
      if (sale_data.customer_id) {
        await customerModel.add_purchase_record(
          sale_data.customer_id,
          result.transaction_id,
          result.total_amount,
          sale_data.payment_method,
          sale_data.items
        );
      }

      res.json({
        success: true,
        message: 'Sale processed successfully',
        data: result
      });
    } else {
      res.status(400).json({
        success: false,
        message: result.error
      });
    }

  } catch (error) {
    console.error('Process sale error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to process sale'
    });
  }
});

// Void a transaction
router.post('/void/:transactionId', authenticateUser, async (req, res) => {
  try {
    const { transactionId } = req.params;
    const { reason } = req.body;

    // Check if user has permission to void transactions
    if (!req.user.permissions?.sales?.includes('void')) {
      return res.status(403).json({
        success: false,
        message: 'Permission denied: Cannot void transactions'
      });
    }

    const result = await salesModel.void_transaction(transactionId, req.user.user_id, reason);

    if (result.success) {
      res.json({
        success: true,
        message: 'Transaction voided successfully'
      });
    } else {
      res.status(400).json({
        success: false,
        message: result.error
      });
    }

  } catch (error) {
    console.error('Void transaction error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to void transaction'
    });
  }
});

// Apply discount to transaction
router.post('/discount/:transactionId', authenticateUser, async (req, res) => {
  try {
    const { transactionId } = req.params;
    const { discount_type, discount_value } = req.body;

    if (!discount_type || discount_value === undefined) {
      return res.status(400).json({
        success: false,
        message: 'Discount type and value are required'
      });
    }

    if (!['percentage', 'fixed'].includes(discount_type)) {
      return res.status(400).json({
        success: false,
        message: 'Discount type must be "percentage" or "fixed"'
      });
    }

    const result = await salesModel.apply_discount(transactionId, discount_type, discount_value);

    if (result.success) {
      res.json({
        success: true,
        message: 'Discount applied successfully',
        data: {
          new_discount: result.new_discount,
          new_total: result.new_total
        }
      });
    } else {
      res.status(400).json({
        success: false,
        message: result.error
      });
    }

  } catch (error) {
    console.error('Apply discount error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to apply discount'
    });
  }
});

// Get sales summary
router.get('/summary', authenticateUser, async (req, res) => {
  try {
    const { start_date, end_date } = req.query;

    if (!start_date || !end_date) {
      return res.status(400).json({
        success: false,
        message: 'Start date and end date are required'
      });
    }

    const result = await salesModel.get_sales_summary(start_date, end_date);

    if (result.success) {
      res.json({
        success: true,
        data: result.summary
      });
    } else {
      res.status(500).json({
        success: false,
        message: result.error
      });
    }

  } catch (error) {
    console.error('Get sales summary error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to get sales summary'
    });
  }
});

// Get recent sales
router.get('/recent', authenticateUser, async (req, res) => {
  try {
    const limit = parseInt(req.query.limit) || 50;
    const offset = parseInt(req.query.offset) || 0;

    // This would require additional implementation in the sales model
    // For now, return a placeholder response
    res.json({
      success: true,
      data: {
        sales: [],
        total: 0,
        limit,
        offset
      }
    });

  } catch (error) {
    console.error('Get recent sales error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to get recent sales'
    });
  }
});

// Get sale by transaction ID
router.get('/transaction/:transactionId', authenticateUser, async (req, res) => {
  try {
    const { transactionId } = req.params;

    // This would require additional implementation in the sales model
    // For now, return a placeholder response
    res.json({
      success: true,
      data: {
        transaction_id: transactionId,
        status: 'completed',
        total_amount: 0,
        items: []
      }
    });

  } catch (error) {
    console.error('Get sale error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to get sale details'
    });
  }
});

// Calculate tax
router.post('/calculate-tax', authenticateUser, async (req, res) => {
  try {
    const { subtotal, tax_rate } = req.body;

    if (subtotal === undefined) {
      return res.status(400).json({
        success: false,
        message: 'Subtotal is required'
      });
    }

    const tax_amount = salesModel.calculate_tax(subtotal, tax_rate || 0.0875);

    res.json({
      success: true,
      data: {
        subtotal: parseFloat(subtotal),
        tax_rate: tax_rate || 0.0875,
        tax_amount: tax_amount,
        total: parseFloat(subtotal) + tax_amount
      }
    });

  } catch (error) {
    console.error('Calculate tax error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to calculate tax'
    });
  }
});

// Get cash drawer status
router.get('/cash-drawer/status', authenticateUser, async (req, res) => {
  try {
    // This would require additional implementation in the sales model
    // For now, return a placeholder response
    res.json({
      success: true,
      data: {
        status: 'closed',
        opening_amount: 0,
        current_amount: 0,
        expected_amount: 0,
        difference: 0,
        opened_at: null,
        closed_at: new Date().toISOString()
      }
    });

  } catch (error) {
    console.error('Get cash drawer status error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to get cash drawer status'
    });
  }
});

// Open cash drawer
router.post('/cash-drawer/open', authenticateUser, async (req, res) => {
  try {
    const { opening_amount } = req.body;

    if (opening_amount === undefined) {
      return res.status(400).json({
        success: false,
        message: 'Opening amount is required'
      });
    }

    // This would require additional implementation in the sales model
    // For now, return a placeholder response
    res.json({
      success: true,
      message: 'Cash drawer opened successfully',
      data: {
        status: 'open',
        opening_amount: parseFloat(opening_amount),
        opened_at: new Date().toISOString()
      }
    });

  } catch (error) {
    console.error('Open cash drawer error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to open cash drawer'
    });
  }
});

// Close cash drawer
router.post('/cash-drawer/close', authenticateUser, async (req, res) => {
  try {
    const { closing_amount } = req.body;

    if (closing_amount === undefined) {
      return res.status(400).json({
        success: false,
        message: 'Closing amount is required'
      });
    }

    // This would require additional implementation in the sales model
    // For now, return a placeholder response
    res.json({
      success: true,
      message: 'Cash drawer closed successfully',
      data: {
        status: 'closed',
        closing_amount: parseFloat(closing_amount),
        closed_at: new Date().toISOString()
      }
    });

  } catch (error) {
    console.error('Close cash drawer error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to close cash drawer'
    });
  }
});

// Get receipt data
router.get('/receipt/:transactionId', authenticateUser, async (req, res) => {
  try {
    const { transactionId } = req.params;

    // This would require additional implementation in the sales model
    // For now, return a placeholder response
    res.json({
      success: true,
      data: {
        transaction_id: transactionId,
        date: new Date().toISOString(),
        items: [],
        subtotal: 0,
        tax_amount: 0,
        discount_amount: 0,
        total_amount: 0,
        payment_method: 'cash',
        employee: req.user.first_name + ' ' + req.user.last_name
      }
    });

  } catch (error) {
    console.error('Get receipt error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to get receipt data'
    });
  }
});

// Print receipt
router.post('/receipt/:transactionId/print', authenticateUser, async (req, res) => {
  try {
    const { transactionId } = req.params;

    // This would integrate with printer service
    // For now, return success
    res.json({
      success: true,
      message: 'Receipt sent to printer successfully'
    });

  } catch (error) {
    console.error('Print receipt error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to print receipt'
    });
  }
});

// Email receipt
router.post('/receipt/:transactionId/email', authenticateUser, async (req, res) => {
  try {
    const { transactionId } = req.params;
    const { customer_email } = req.body;

    if (!customer_email) {
      return res.status(400).json({
        success: false,
        message: 'Customer email is required'
      });
    }

    // This would integrate with email service
    // For now, return success
    res.json({
      success: true,
      message: 'Receipt email sent successfully'
    });

  } catch (error) {
    console.error('Email receipt error:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to email receipt'
    });
  }
});

module.exports = router;